python/typemap_utils.cpp

```diff
@@ -0,0 +1,427 @@
+    }
+
+    double radius = cyl.subclass.cylinder_data->radius;
+    double height = cyl.subclass.cylinder_data->height;
+    vector3 axis = cyl.subclass.cylinder_data->axis;
+
+    *wedge = make_wedge(cyl.material, cyl.center, radius, height, axis, wedge_angle, wedge_start);
+
+    geometric_object_destroy(cyl);
+
+    return 1;
+}
+
+static int pycone_to_cone(PyObject *py_cone, geometric_object *cone) {
+    geometric_object cyl;
+    if(!pycylinder_to_cylinder(py_cone, &cyl)) {
+        return 0;
+    }
+
+    double radius2;
+    if(!get_attr_dbl(py_cone, &radius2, "radius2")) {
+        cone->subclass.cylinder_data = NULL;
+        geometric_object_destroy(cyl);
+        return 0;
+    }
+
+    double radius = cyl.subclass.cylinder_data->radius;
+    double height = cyl.subclass.cylinder_data->height;
+    vector3 axis = cyl.subclass.cylinder_data->axis;
+
+    *cone = make_cone(cyl.material, cyl.center, radius, height, axis, radius2);
+
+    geometric_object_destroy(cyl);
+
+    return 1;
+}
+
+static int pyblock_to_block(PyObject *py_blk, geometric_object *blk) {
+    material_type material;
+    vector3 center, e1, e2, e3, size;
+
+    if(!get_attr_material(py_blk, &material) ||
+       !get_attr_v3(py_blk, &center, "center") ||
+       !get_attr_v3(py_blk, &e1, "e1") ||
+       !get_attr_v3(py_blk, &e2, "e2") ||
+       !get_attr_v3(py_blk, &e3, "e3") ||
+       !get_attr_v3(py_blk, &size, "size")) {
+
+        blk->subclass.block_data = NULL;
+        return 0;
+    }
+
+    *blk = make_block(material, center, e1, e2, e3, size);
+    return 1;
+}
+
+static int pyellipsoid_to_ellipsoid(PyObject *py_ell, geometric_object *e) {
+    geometric_object blk;
+    if(!pyblock_to_block(py_ell, &blk)) {
+        return 0;
+    }
+
+    material_type material = blk.material;
+    vector3 center = blk.center;
+    vector3 e1 = blk.subclass.block_data->e1;
+    vector3 e2 = blk.subclass.block_data->e2;
+    vector3 e3 = blk.subclass.block_data->e3;
+    vector3 size = blk.subclass.block_data->size;
+
+    *e = make_ellipsoid(material, center, e1, e2, e3, size);
+
+    geometric_object_destroy(blk);
+
+    return 1;
+}
+
+static int py_list_to_gobj_list(PyObject *po, geometric_object_list *l) {
+    if(!PyList_Check(po)) {
+        PyErr_SetString(PyExc_TypeError, "Expected a list");
+        return 0;
+    }
+
+    int length = PyList_Size(po);
+
+    l->num_items = length;
+    l->items = new geometric_object[length];
+
+    for(int i = 0; i < length; i++) {
+        PyObject *py_gobj = PyList_GetItem(po, i);
+        geometric_object go;
+        if(!py_gobj_to_gobj(py_gobj, &go)) {
+            return 0;
+        }
+        geometric_object_copy(&go, &l->items[i]);
+    }
+
+    return 1;
+}
+
+static std::string py_class_name_as_string(PyObject *po) {
+    PyObject *py_type = PyObject_Type(po);
+    PyObject *name = PyObject_GetAttrString(py_type, "__name__");
+
+    char *bytes = PyObject_ToCharPtr(name);
+
+    std::string class_name(bytes);
```

Comment:
I tried freeing `bytes` and saw a crash, then I noticed that the [documentation](https://docs.python.org/2/c-api/string.html#c.PyString_AsString) says this memory "must not be deallocated." It's shared with the interpreter, so python handles it.